### DISCUSSION TITLE
Post-hook of block import in client

### DISCUSSION THREAD

**Discussion by liuchengxu** [paritytech/substrate#13927] [Ideas]
Substrate provides the feature of subscribing to each imported block event which is typically considered as the completion of importing a block if I'm not wrong. However, we have the concept of domain block at Subspace which needs to run some special post-hook logics (generate and store the receipt of the imported block, etc) after the block is imported (by calling `client.import_block()`) which imposes a challenge: when the imported block notification from the client is out, the entire domain block import pipeline is not yet finished in theory due to the special post-hook logics, which is mostly okay in practice but I have encountered some hassles in the test environment.

```Rust
#[test]
async fn test_domain() {
    // Wait for one block based on the imported block notification from client.
    alice.wait_for_blocks(1).await;

    // Test some logic that involves the receipt of the last imported block (done in the post-hook logic).
    // Since the post-hook of the last imported block may be unfinished, the receipt is unavailable at
    // this point, causing the test flaky due to this uncertainty.
}
```

In short, I'm wondering if can we add a hook at a point which is after the block is imported into the database and before sending the notification of block imported so that the utility of BlockchainEvents can still be seamlessly used and generally make Substrate more flexible.

https://github.com/paritytech/substrate/blob/9a07e9d564636198afef9260de390ebcd40ac1fc/client/service/src/client/client.rs#L330
